File: backend/app/schemas/meeting_import.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator

from app.models.kb import MeetingImportStatus


class MeetingImportConfig(BaseModel):
    max_upload_bytes: int
    allowed_extensions: list[str]
    allowed_mime_types: list[str]
    # Kept as a compatibility alias for early clients.
    mime_types: dict[str, list[str]]
    statuses: list[MeetingImportStatus]
    supports_existing_document: bool = True


class MeetingImportRead(BaseModel):
    import_id: str
    org_id: str
    kb_id: str
    organization_id: str
    knowledge_base_id: str
    document_id: str
    file: dict[str, Any]
    status: MeetingImportStatus
    current_step: str
    progress: int
    failure: dict[str, Any] | None = None
    can_retry: bool
    created_at: datetime
    updated_at: datetime
    metadata: dict[str, Any]


class BlockEdit(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=False)
    block_id: str = Field(min_length=1, max_length=100)
    text: str | None = None
    block_type: str | None = None
    heading_path: list[Any] | None = None
    table_markdown: str | None = None


class RevisionPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_version: int = Field(ge=1)
    block_edits: list[BlockEdit] = Field(default_factory=list)


class FindRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    query: str = Field(min_length=1)
    scope: Literal["FULL", "BLOCK"] = "FULL"
    case_sensitive: bool = False
    block_id: str | None = None


class ReplaceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    mode: Literal["CURRENT", "ALL"] = "CURRENT"
    query: str = Field(min_length=1)
    replacement: str
    scope: Literal["FULL", "BLOCK"] = "FULL"
    case_sensitive: bool = False
    expected_version: int = Field(ge=1)
    block_id: str | None = None
    match_index: int | None = Field(default=None, ge=0)
    preview: bool = False


class UndoRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_version: int = Field(ge=1)


class MeetingMetadataPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_version: int = Field(ge=1)
    title: str | None = Field(default=None, max_length=255)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    location: str | None = Field(default=None, max_length=500)
    online_url: HttpUrl | None = None
    organizer: str | None = Field(default=None, max_length=255)
    topic: str | None = Field(default=None, max_length=255)
    description: str | None = None
    meeting_purpose: str | None = None
    discussion_topics: str | None = None
    meeting_date: str | None = None
    advisor_selection_criteria: str | None = None
    advisor_names: str | None = None
    internal_attendees: str | None = None
    recorder: str | None = None

    @field_validator("starts_at", "ends_at")
    @classmethod
    def require_timezone(cls, value: datetime | None) -> datetime | None:
        if value is not None and (value.tzinfo is None or value.utcoffset() is None):
            raise ValueError("时间必须携带时区信息")
        return value


class ConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_version: int = Field(ge=1)
    expected_metadata_version: int = Field(ge=1)
    title: str | None = Field(default=None, max_length=255)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    location: str | None = Field(default=None, max_length=500)
    online_url: HttpUrl | None = None
    organizer: str | None = Field(default=None, max_length=255)
    topic: str | None = Field(default=None, max_length=255)
    description: str | None = None
    meeting_purpose: str | None = None
    discussion_topics: str | None = None
    meeting_date: str | None = None
    advisor_selection_criteria: str | None = None
    advisor_names: str | None = None
    internal_attendees: str | None = None
    recorder: str | None = None

    @field_validator("starts_at", "ends_at")
    @classmethod
    def require_timezone(cls, value: datetime | None) -> datetime | None:
        if value is not None and (value.tzinfo is None or value.utcoffset() is None):
            raise ValueError("时间必须携带时区信息")
        return value


class SourceRefRead(BaseModel):
    block_id: str
    page_number: int | None = None
    slide_number: int | None = None
    speaker: str | None = None
    start_ms: int | None = None
    end_ms: int | None = None


class RevisionBlockRead(BaseModel):
    block_id: str
    block_type: str
    order: int
    text: str
    heading_path: list[Any] = Field(default_factory=list)
    table_markdown: str | None = None
    source_ref: SourceRefRead | None = None


class RevisionRead(BaseModel):
    revision_id: UUID
    document_id: UUID
    version: int
    revision_number: int
    status: str
    created_by: UUID
    blocks: list[RevisionBlockRead]
    created_at: datetime
    confirmed_at: datetime | None = None
    confirmed_by: UUID | None = None


class MeetingMetadataField(BaseModel):
    value: Any = None
    suggested_value: Any = None
    confidence: float | str | None = None
    confidence_label: str | None = None
    source: list[dict[str, Any]] = Field(default_factory=list)
    needs_confirmation: bool = False
    user_modified: bool = False


class ReviewRead(BaseModel):
    import_id: UUID
    organization_id: UUID
    knowledge_base_id: UUID
    document: dict[str, Any]
    file: dict[str, Any]
    status: str
    meeting_id: UUID | None = None
    original_blocks: list[RevisionBlockRead]
    current_revision: RevisionRead | None = None
    revision_history: list[RevisionRead] = Field(default_factory=list)
    metadata: dict[str, MeetingMetadataField]
    metadata_version: int
    needs_confirmation_count: int
    vectorization: "VectorizationRead"


class VectorizationRead(BaseModel):
    job_id: str | None = None
    status: Literal["PENDING", "RUNNING", "SYNCED", "STALE", "FAILED"]
    revision_id: UUID | None = None
    current_revision_version: int | None = None
    vectorized_revision_version: int | None = None
    current_node: str | None = None
    progress: int = 0
    error_code: str | None = None
    error_message: str | None = None
    error: dict[str, Any] | None = None
    retryable: bool = False


class VectorizationReadResponse(VectorizationRead):
    import_id: UUID


class VectorizeRequest(BaseModel):
    expected_version: int = Field(ge=1)


class ConfirmRead(BaseModel):
    meeting_id: UUID
    import_id: UUID
    revision_id: UUID
    status: str
    rag_job_id: str | None = None
    rag_status: str | None = None
    rag_error: str | None = None
    rag_retryable: bool = False
    ai_task_id: UUID | None = None
    question_generation_status: str | None = None
    meeting_status: str = "QUESTION_GENERATION_QUEUED"

File: frontend/src/types/meetingImport.ts
export type MeetingImportStatus =
  | 'UPLOADING'
  | 'UPLOADED'
  | 'VALIDATING'
  | 'PARSING'
  | 'EXTRACTING_METADATA'
  | 'READY_FOR_REVIEW'
  | 'FAILED'
  | 'CANCELLED'
  | 'CANCELED'
  | (string & {})

export interface MeetingImportConfig {
  max_upload_bytes: number
  allowed_extensions: string[]
  allowed_mime_types: string[]
  mime_types?: Record<string, string[]>
}

export interface MeetingImportRead {
  id: string
  import_id?: string
  knowledge_base_id: string
  document_id?: string
  filename?: string
  status: MeetingImportStatus
  progress?: number
  progress_percent?: number
  current_step?: string
  error_code?: string
  error_message?: string
  can_retry?: boolean
  created_at?: string
  updated_at?: string
  [key: string]: unknown
}

export type ReviewBlockType = 'paragraph' | 'speaker' | 'heading' | 'table' | 'unknown' | (string & {})

export interface ReviewBlock {
  id: string
  type?: ReviewBlockType
  block_type?: ReviewBlockType
  text: string
  table_markdown?: string | null
  speaker?: string | null
  start_ms?: number | null
  end_ms?: number | null
  page_number?: number | null
  paragraph_number?: number | null
  source?: Record<string, unknown> | null
  [key: string]: unknown
}

export interface ReviewMetadataField {
  value: string | null
  suggested_value?: string | null
  confidence?: number | null
  confidence_label?: string | null
  source?: { block_id?: string; quote?: string; page_number?: number; [key: string]: unknown } | Array<{ block_id?: string; quote?: string; page_number?: number; [key: string]: unknown }> | null
  needs_confirmation?: boolean
  user_modified?: boolean
}

export interface ReviewMetadata {
  title: ReviewMetadataField
  starts_at: ReviewMetadataField
  ends_at: ReviewMetadataField
  location: ReviewMetadataField
  online_url: ReviewMetadataField
  organizer: ReviewMetadataField
  topic: ReviewMetadataField
  description: ReviewMetadataField
  meeting_purpose?: ReviewMetadataField
  discussion_topics?: ReviewMetadataField
  meeting_date?: ReviewMetadataField
  advisor_selection_criteria?: ReviewMetadataField
  advisor_names?: ReviewMetadataField
  internal_attendees?: ReviewMetadataField
  recorder?: ReviewMetadataField
  [key: string]: ReviewMetadataField
}

export interface ReviewRevision {
  id: string
  revision_number: number
  status?: string
  version: number
  blocks: ReviewBlock[]
  created_at?: string
  updated_at?: string
  created_by?: string
  creator?: string
  [key: string]: unknown
}

export type MeetingImportVectorizationStatus = 'PENDING' | 'RUNNING' | 'SYNCED' | 'STALE' | 'FAILED' | (string & {})

export interface MeetingImportVectorization {
  status: MeetingImportVectorizationStatus
  job_id?: string | null
  current_node?: string | null
  progress?: number | null
  error_code?: string | null
  error_message?: string | null
  retryable?: boolean
  current_revision_version?: number | null
  vectorized_revision_version?: number | null
  [key: string]: unknown
}

export interface MeetingImportReview {
  import: MeetingImportRead
  file?: { filename?: string; [key: string]: unknown }
  status: MeetingImportStatus | string
  document?: Record<string, unknown>
  knowledge_base?: Record<string, unknown>
  meeting_id?: string | null
  original_blocks: ReviewBlock[]
  current_revision: ReviewRevision | null
  revisions: ReviewRevision[]
  meeting_metadata: ReviewMetadata
  metadata_version: number
  needs_confirmation_count: number
  vectorization: MeetingImportVectorization
  [key: string]: unknown
}

export interface DuplicateImportDetails {
  existing_document_id?: string
  existing_import_id?: string
  [key: string]: unknown
}
